lib/cdo/config.rb

```diff
@@ -4,31 +4,66 @@
 
 module Cdo
   # Loads and combines structured application configuration settings.
-  class Config < OpenStruct
-    # Soft-freeze: Don't allow any config items to be created/modified,
-    # but allow stubbing for unit tests.
-    def freeze
-      # Some implementation changes in OpenStruct in Ruby 3.0 affect behavior
-      # that this class is relying on, so we need to include some special logic
-      # here. See https://bugs.ruby-lang.org/issues/15409#note-9
-      @table.each_key {|k| new_ostruct_member!(k.to_sym)}
+  class Config
+    def initialize
+      @table = {}
+    end
+
+    # Converts the Config object to a hash
+    def to_h
+      @table.dup
+    end
+
+    # Computes a hash code for this Config object.
+    def hash
+      @table.hash
+    end
+
+    # Returns the value of an attribute, or `nil` if there is no such attribute.
+    def [](name)
+      @table[name.to_sym]
+    end
+
+    # Returns `true` if the given name is a configuration key
+    def key?(name)
+      @table.key?(name.to_sym)
+    end
+
+    # Implement our own soft-freeze: Don't allow any config items to be
+    # created/modified, but allow stubbing for unit tests.
+    def freeze_config
       @frozen = true
     end
 
-    def method_missing(key, *args)
-      return self[key.to_sym] if args.empty? && @table.key?(key.to_sym) # accommodate https://bugs.ruby-lang.org/issues/15409#note-9
-      raise ArgumentError, "Undefined #{self.class} reference: #{key}", caller(1) if @frozen
-      super
+    # Inspired by OpenStruct; if unfrozen, allow assigning new configuration
+    # values with `=` and default undefined values to nil
+    # See https://github.com/ruby/ostruct/blob/e61b4464a033c38a65657eaf0467d12e2c7b9ec1/lib/ostruct.rb#L207-L229
+    def method_missing(mid, *args)
```

Comment:
Could we use `ActiveSupport::OrderedOptions`? Looks like it also overrides `method_missing` to provide the functionality we need, and it sub-classes `Hash`.

Reply:
In theory I think it'd probably be possible, but would require much more extensive refactoring than this approach. In particular, our [dynamic configuration rerendering logic as it's currently implemented requires us to be able to cache and restore the underlying data store between each interation](https://github.com/code-dot-org/code-dot-org/blob/406c195945c2571da5b4670dd6a9654fab5f55ac/lib/cdo/config.rb#L34-L37), and I have no idea how we'd go about doing that with something that subclasses rather than wraps its data store. I'd also be worried about the [couple](https://github.com/code-dot-org/code-dot-org/blob/406c195945c2571da5b4670dd6a9654fab5f55ac/lib/cdo/secrets_config.rb#L70) of [instances](https://github.com/code-dot-org/code-dot-org/blob/406c195945c2571da5b4670dd6a9654fab5f55ac/lib/cdo/secrets_config.rb#L92) of full-table iteration that we currently do in the secrets config.

I'd prefer to stick with something that maintains existing functionality for now, rather than embark on a full rewrite of this class